show where JavaScript files should be put (after comments)  [ci skip] (#42328)
rails/rails: guides/source/working_with_javascript_in_rails.md

means that it'll get downloaded on the first page load and then be cached on
every page after that. Lots of little benefits really add up.

> As usual, Rails has opinions about how the software should be structured and this applies to JavaScript too. Rails provides an `app\javascript` directory as a home for our JavaScript. If we use more and more JavaScript in our app, it is worth spreading out the files across a directory structure, but for now, it is better to split the JavaScript and put it in two places. The function itself should go in `app\javascript\paint.js` after adding the export statements

Reviewer: Paths are commonly written with forward slashes in the documentation.

```suggestion
As usual, Rails has opinions about how the software should be structured and this applies to JavaScript too. Rails provides an `app/javascript` directory as a home for our JavaScript. If we use more and more JavaScript in our app, it is worth spreading out the files across a directory structure, but for now, it is better to split the JavaScript and put it in two places. The function itself should go in `app/javascript/paint.js` after adding the export statements.
```

Reviewer: Perhaps the example could point towards a file inside an `app/javascript/components/` directory?

This could help structure the first JavaScript files in the app and simplify this text down to something like this:

```suggestion
As usual, Rails has opinions about how software should be structured and this applies to JavaScript too. The function would go for example in `app/javascript/components/paint.js`, with the following export statement.
```

```js
export default function paintIt(element, backgroundColor, textColor) {
  element.style.backgroundColor = backgroundColor;
  if (textColor) {
    element.style.color = textColor;
  }
}
```

> and the listener should be added to `app\javascript\packs\application.js`.

Reviewer: ```suggestion

And the listener should be added to `app/javascript/packs/application.js`.
```

```js
// This file is automatically compiled by Webpack, along with any other files
// present in this directory. You're encouraged to place your actual application logic in
// a relevant structure within app/javascript and only use these pack files to reference
// that code so it'll be compiled.

import Rails from "@rails/ujs"
import Turbolinks from "turbolinks"
import * as ActiveStorage from "@rails/activestorage"
import "channels"
import paintIt from "paint"
 

Rails.start()
Turbolinks.start()
ActiveStorage.start()

> window.addEventListener("load", () => {

Reviewer: Ideally here, wouldn’t we just have `window.addEventListener("load", initPaint)` instead, in order not to bloat the `application.js`?

Having an extra function around the whole initializing code seems to me the better practice.

  const links = document.querySelectorAll(
    "a[data-background-color]"
  );
  links.forEach((element) => {
    element.addEventListener("click", (event) => {
      event.preventDefault();

      const {backgroundColor, textColor} = element.dataset;
      paintIt(element, backgroundColor, textColor);
    });
  });
});
```

> One final change (see below) is to make the listener for a (slightly) different event. As the default Rails app uses Turbolinks, we don't get a "load" event every time. Mostly, we need to listen for "turbolinks:load" instead.

Reviewer: If I may suggest the following minor changes, I feel that they help simplify the text:

```suggestion
Since default Rails app uses Turbolinks ([see below](#turbolinks)), one final change would be to listen for the `turbolinks:load` event instead of `load`.
```
